File: June 2025/30.cpp
class Solution {
  public:
    int maxMinHeight(vector<int> &arr, int k, int w) {
        int n = arr.size();

        auto canReach = [&](int target) {
            vector<int> diff(n + 1, 0);
            int water = 0, curr_add = 0;

            for (int i = 0; i < n; ++i) {
                curr_add += diff[i];
                int effective_height = arr[i] + curr_add;

                if (effective_height < target) {
                    int needed = target - effective_height;
                    water += needed;
                    if (water > k) return false;

                    curr_add += needed;
                    if (i + w < n + 1)
                        diff[i + w] -= needed;
                }
            }
            return true;
        };

        int low = *min_element(arr.begin(), arr.end());
        int high = *max_element(arr.begin(), arr.end()) + k;
        int ans = low;

        while (low <= high) {
            int mid = (low + high) / 2;
            if (canReach(mid)) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return ans;
    }
};

File: June 2025/30.py
class Solution():
    def maxMinHeight(self, arr, k, w):
        n = len(arr)

        def canReach(target):
            water = 0
            diff = [0] * (n + 1)
            curr_add = 0

            for i in range(n):
                curr_add += diff[i]
                effective_height = arr[i] + curr_add

                if effective_height < target:
                    needed = target - effective_height
                    water += needed
                    if water > k:
                        return False
                    curr_add += needed
                    if i + w < len(diff):
                        diff[i + w] -= needed
            return True

        low, high = min(arr), max(arr) + k
        ans = low

        while low <= high:
            mid = (low + high) // 2
            if canReach(mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1

        return ans
